### DISCUSSION TITLE
Inconsistency between UI and data when using Zustand and NextJs

### DISCUSSION THREAD

**Discussion by andrey-sheremeta** [pmndrs/zustand#1943] [Q&A]
Hello.

I faced one problem while using NextJs(12) and Zustand(4). There is an inconsistency between UI and data, leading to many different defects. 

F.e.
you have two pages A & B; each page returns its data via `getServerSideProps`. so on the client side when a user navigates from page A to page B, to properly render a page we need to update Zustand store with new data received from the server. The current approach that is described here https://docs.pmnd.rs/zustand/guides/initialize-state-with-props encourages to use `useEffect` 
![image](https://github.com/pmndrs/zustand/assets/70937584/1e9f15f9-a665-43fd-83bd-cb8242ef4088) but it leads to different defects (addition rerendering, inconsistency between UI and data and etc.). Is there any alternative way to update the store, without using useEffect?

In my understanding, it's quite a critical thing.

repo with example - https://github.com/andrey-sheremeta/zustand-state-update

I was trying to remove `useEffect` it seems work but then only react returns an error - https://github.com/facebook/react/issues/18178#issuecomment-595846312

**Comment by noconsulate** [comment]
I just went through all this but I'm using the App router rather than the Pages router, so I'm using server and client components. I got it to work doing [this](https://vercel.com/guides/react-context-state-management-nextjs) though I had to also wrap the client component provider component in another server component that does the initial fetch. That might not make any sense if you're not used to using the App router.

You may find [this](https://github.com/vercel/next.js/discussions/45763) helpful, as they seem to be doing what you're after using getServerSideProps and all that. 

It's worth noting once I got the initial state with props and context providing and all that my store and UI are working flawlessly.

**Reply by andrey-sheremeta** [reply]
hi @noconsulate. Thanks for your time and efforts. Unfortunately, [this ](https://github.com/vercel/next.js/discussions/45763)did not help. The suggested approach also depends on the `useLayouteffect` that we need to avoid.  

![image](https://github.com/pmndrs/zustand/assets/70937584/7bf65f25-9318-4c50-ace4-b8ecbe49d1c0)

**Reply by Harshit-Mourya** [reply]
Hi @andrey-sheremeta 

I came across this issue and found it really interesting — especially the part about Zustand and `getServerSideProps`.

I'm new to open source and this would be my first contribution, so I’d like to try and help if the issue is still open.

Have there been any updates or suggestions since the last comment?

**Reply by andrey-sheremeta** [reply]
hey @Harshit-Mourya, thanks for being interested in this issue. Yes, this problem is still relevant. In our application, we found a workaround that partially works, but it doesn't resolve all problems. So, if you find a solution, the native by nature will be perfect, I think.

**Reply by dbritto-dev** [reply]
@andrey-sheremeta did you tried with `store.setState(state => ({ ...state, ...initProps }), true)`? btw, I'd use createInitState function instead of getting that from the store itself

**Reply by Harshit-Mourya** [reply]
Thanks @andrey-sheremeta  and @dbritto-dev for your inputs!

I’ve actually submitted a PR here that tries to address this issue:
[fix: sync Zustand store with SSR props after client navigation](https://github.com/andrey-sheremeta/zustand-state-update/pull/1)

It introduces a `useSyncPageName` hook to safely hydrate the store during SSR and client navigation without extra re-renders or stale state.

It’d be great to get feedback or suggestions for improvements.

**Reply by dbritto-dev** [reply]
@Harshit-Mourya btw, I wouldn't use global store with next.js we recommend to create them on each request

**Reply by Harshit-Mourya** [reply]
Thanks for the tip! Makes sense — I’ll try switching to a per-request store pattern instead of a global one.

Just to clarify: would wrapping `createStore` in a function like `initStore()` and calling that in each request handler (or component tree for RSC) be the right approach?

**Reply by dbritto-dev** [reply]
@Harshit-Mourya here you go -> https://stackblitz.com/edit/nextjs-and-zustand-persist-snefbl8b?file=README.md

**Reply by Harshit-Mourya** [reply]
Updated the PR to use a per-request store with `createUiStore()` as suggested.  
Also refactored `_app.tsx` and added `UiStoreContext` to provide the store via context.  

PR: https://github.com/andrey-sheremeta/zustand-state-update/pull/1
